packages/protocol/contracts/L1/libs/LibVerifying.sol

```diff
@@ -237,24 +233,24 @@ library LibVerifying {
             }
 
             if (numBlocksVerified > 0) {
-                uint64 lastVerifiedBlockId = b.lastVerifiedBlockId + numBlocksVerified;
-
-                // Update protocol level state variables
-                state.slotB.lastVerifiedBlockId = lastVerifiedBlockId;
-
-                // Store the L2's state root as a signal to the local signal
-                // service to allow for multi-hop bridging.
-                //
-                // This also means if we verified more than one block, only the last one's stateRoot
-                // is sent as a signal and verifiable with merkle proofs, all other blocks'
-                // stateRoot are not.
-                ISignalService(resolver.resolve("signal_service", false)).relayChainData(
-                    config.chainId, LibSignals.STATE_ROOT, stateRoot
-                );
-
-                emit CrossChainSynced(
-                    uint64(block.number), lastVerifiedBlockId, blockHash, stateRoot
-                );
+                b.lastVerifiedBlockId += numBlocksVerified;
+
+                if (config.maxUnrelayedBlocks > 0) {
+                    uint256 unrelayedBlocks = numBlocksVerified + b.unrelayedBlocks;
+                    if (unrelayedBlocks < config.maxUnrelayedBlocks) {
+                        // config.maxUnrelayedBlocks is uint8
+                        b.unrelayedBlocks = uint8(unrelayedBlocks);
+                    } else {
+                        // Store the L2's state root as a signal to the local signal
+                        // service to allow for multi-hop bridging.
+                        ISignalService(resolver.resolve("signal_service", false)).relayChainData(
+                            config.chainId, b.lastVerifiedBlockId, LibSignals.STATE_ROOT, stateRoot
+                        );
+                        b.unrelayedBlocks = 0;
+                    }
+                }
```

Comment:
I think could be made a bit more efficient if we just query the latest synced state from the signal service
```
ISignalService signalService = ISignalService(resolver.resolve("signal_service", false));
uint64 lastSyncedBlockId = signalService.topBlockId(config.chainId, LibSignals.STATE_ROOT);
if (b.lastVerifiedBlockId > lastSyncedBlockId + config.maxUnrelayedBlocks) {
    signalService.relayChainData(config.chainId, b.lastVerifiedBlockId, LibSignals.STATE_ROOT, stateRoot);
}
```

Reading the data is more expensive (but also makes the call to `relayChainData` cheaper if it's needed because the signal service contract is already warm), but no need to update `b.unrelayedBlocks` which is even more expensive.